src/librustc_mir/dataflow/impls/mod.rs

```diff
@@ -390,33 +395,34 @@ impl<'a, 'gcx, 'tcx> BitDenotation for MaybeUninitializedLvals<'a, 'gcx, 'tcx> {
         for e in sets.on_entry.words_mut() { *e = !0; }
 
         drop_flag_effects_for_function_entry(
-            self.tcx, self.mir, self.mdpe,
+            self.mir, &self.mdpe.move_data,
+            |lv| place_contents_drop_state_cannot_differ(self.tcx, self.mir, lv),
             |path, s| {
                 assert!(s == DropFlagState::Present);
                 sets.on_entry.remove(&path);
-            });
+            })
     }
 
     fn statement_effect(&self,
                         sets: &mut BlockSets<MovePathIndex>,
                         location: Location)
     {
         drop_flag_effects_for_location(
-            self.tcx, self.mir, self.mdpe,
+            &self.mdpe.move_data,
             location,
-            |path, s| Self::update_bits(sets, path, s)
-        )
+            |lv| place_contents_drop_state_cannot_differ(self.tcx, self.mir, lv),
+            |path, s| Self::update_bits(sets, path, s))
     }
 
     fn terminator_effect(&self,
                          sets: &mut BlockSets<MovePathIndex>,
                          location: Location)
     {
         drop_flag_effects_for_location(
-            self.tcx, self.mir, self.mdpe,
+            &self.mdpe.move_data,
             location,
-            |path, s| Self::update_bits(sets, path, s)
-        )
+            |lv| place_contents_drop_state_cannot_differ(self.tcx, self.mir, lv),
```

Comment:
One advantage to this approach is that its trivial to add instrumentation to the individual closures. Also, you need both the `tcx` and the `mir`, so to get the full effect here (in terms of avoiding having to thread either of those down) I'd have to at the very least make a struct that held both of them, rather than just passing either on its own.

Having said that, I *was* considering moving to a trait instead of arbitrary closures, largely because I wanted a single thing to carry both of the closures that I occasionally pass in. So I'll look into that.

Reply:
(Oh, maybe you mean make it a method on `tcx` that takes the `mir` as an argument, and then pass in a closure that does `|arg| self.tcx.new_method(self.mir, arg)` ...? Not sure how that's better/clearer that `|arg| current_func(self.tcx, self.mir, arg)`, though...)